### PR TITLE
{SYCL][PI][L0] - Eliminate std::string construction/destruction overh…

### PR DESCRIPTION
**Pull request by kbsmith-intel**
…ead.

This change eliminates std::string construction/destruction overhead
for calls to ZeCall::doCall, and similar entry points. The changes have
no observable functional effect but reduce the overall size of the
built library by 8-9%, and eliminate the memory allocation and
std::string construction ond destruction overhead, thus lowering
overall execution time on every level zero API call that uses the
macros ZE_CALL and ZE_CALL_NOCHECK.

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -456,18 +456,18 @@ inline void zeParseError(ze_result_t ZeError, std::string &ErrorString) {
...
+                           const char *ZeArgs, bool TraceError) {
+  zePrint("ZE ---> %s%s\n", ZeName, ZeArgs);
 
   if (ZeDebug & ZE_DEBUG_CALL_COUNT) {
     ++(*ZeCallCount)[ZeName];
   }
 
   if (ZeResult && TraceError) {
-    std::string ErrorString;
+    const char *ErrorString = "Unknown";
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:468]: Why did you need to do "Unknown" instead of "" before? BTW, I assume this value is always unused, i.e. `zeParseError` always updates the value, right?

**Review by smaslov-intel — APPROVED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (kbsmith-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:468]: I chose unknown because there is a case where zeParseError has a default in the case, and doesn't set ErrorString.  I can change it to "", but I thought "Unknown" worked nicest.  I think it is unimportant, and in practice will be completely unused.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:468]: OK

**Comment by keryell** [comment]
That looks like a good motivation.
But I wonder whether this is not a good use case for some `std::string_view`...

**Comment by AGindinson** [comment]
> That looks like a good motivation.
> But I wonder whether this is not a good use case for some `std::string_view`...

AFAIU, C++17 features are not available in the current library/runtime build.

**Comment by kbsmith-intel** [comment]
> > That looks like a good motivation.
> > But I wonder whether this is not a good use case for some `std::string_view`...
> 
> AFAIU, C++17 features are not available in the current library/runtime build.

I did a quick check where I used std::string_view instead of the const char * for the params of doCall.  This is almost as good with respect to the overhead as the const char *, but not quite as good.  In particular, for every usage of the macro ZE_CALL or ZE_CALL_NOCHECK, with string_view, there is extra overhead for construction of the string_view.  With const char *, it just loads the address of the string into a register (for each of the two string parameters), while for string_view ir loads both the address of the string, and the strings length.  So, this causes a bit more overhead, and this really is unnecessary since all uses in doCall really just use the string itself.

So, my preference is to keep this implementation using const char *.
